### PR TITLE
azure: StrictCacheUpdates to disable proactive vmss cache updates

### LINKED ISSUE #7432: CAS appears to lose track of instances in VMSS and makes incorrect scaling decisions
**Which component are you using?**:
cluster-autoscaler

**What version of the component are you using?**:

Component version: v1.28.6

**What k8s version are you using (`kubectl version`)?**:

`kubectl version` output may not be accurate, since our data is from a month ago. Instead, I've pulled the cluster version from the logs: `v1.29.6`

**What environment is this in?**:
Azure, manually running a k8s cluster on top of VMSSes (not AKS).

**What did you expect to happen?**:
We expected the cluster autoscaler to scale a VMSS up to a size larger than it currently was.

**What happened instead?**:
The cluster autoscaler mistakenly resized a VMSS from 86 nodes down to 53, based on an erroneous in-memory value showing that the VMSS had 48 nodes.

**How to reproduce it (as minimally and precisely as possible)**:
Reproing this has been really difficult. We're not sure exactly what causes this to occur. Each time we've seen it happen, it has been preceded by errors from the Azure API. Specifically, the error rate for `vmss_wait_for_delete_instances_result` calls increases to 100%, with the CAS reporting node deletion timeouts along with other errors.

**Anything else we need to know?**:

We saw some errors like this prior to the scale-down event:
```
E1003 22:15:28.247174       1 azure_scale_set.go:200] Failed to update the capacity for vmss <elided> with error Code="CannotAllocateRemainingVMsInAvailabilitySet" Message="Delete/Deallocate operation on VM 'vmss-<elided>_679' failed because the remaining VMs in the Availability Set '_vmss-<elided>_<elided>b3' cannot be allocated together. Changes in Availability Set allocation need to be executed atomically. Please deallocate or delete some or all of these VMs before retrying the current operation. \r\nDetails: Allocation failed. If you are trying to add a new VM to a Virtual Machine Scale Set with a single placement group or update/resize an existing VM in a Virtual Machine Scale Set with a single placement group, please note that such allocation is scoped to a single cluster, and it is possible that the cluster is out of capacity. Please read more about improving likelihood of allocation success at http://aka.ms/allocation-guidance." Target="679", invalidate the cache so as to get the real size from API
```
The CAS _really_ wanted to delete this instance (along with 3-4 others). It kept trying to delete this instance over and over and over again.

We also saw the in-memory node count for the VMSS steadily decrease, and this decrease seems to be correlated with the reduction:
```
I1003 22:31:58.325188       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 93
I1003 22:32:51.759222       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 88
I1003 22:33:53.728006       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 82
I1003 22:34:50.949732       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 78
I1003 22:35:58.258156       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 72
I1003 22:36:57.201547       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 67
I1003 22:37:54.935086       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 61
I1003 22:38:57.801178       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 57
I1003 22:39:43.711224       1 azure_scale_set.go:149] VMSS: vmss-<elided> returning in-memory size: 53
```

<details>

<summary>Here are the relevant logs bracketing that last line.</summary>

```
I1003 22:39:43.260997       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 53
I1003 22:39:43.261054       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 53
I1003 22:39:43.710852       1 clusterstate.go:1094] Failed adding 1 nodes (1 unseen previously) to group vmss-<elided> due to OutOfResource.provisioning-state-failed; errorMessages=[]string{"Azure failed to provision a node for this node group"}
I1003 22:39:43.711211       1 static_autoscaler.go:854] Deleting 1 from vmss-<elided> node group because of create errors
# original log line from above
I1003 22:39:43.711224       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 53
I1003 22:39:43.711258       1 azure_scale_set.go:376] Deleting vmss instances [azure:///subscriptions/<elided>/resourceGroups/<elided>/providers/Microsoft.Compute/virtualMachineScaleSets/vmss-<elided>/virtualMachines/679]
I1003 22:39:43.711315       1 azure_scale_set.go:426] Calling virtualMachineScaleSetsClient.DeleteInstancesAsync(&[679])
I1003 22:39:43.894553       1 static_autoscaler.go:439] Some nodes that failed to create were removed, skipping iteration
I1003 22:39:43.894587       1 azure_scale_set.go:184] Calling virtualMachineScaleSetsClient.WaitForDeleteInstancesResult(&[679]) for vmss-<elided>
I1003 22:39:54.302812       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 52
I1003 22:39:55.477245       1 node_instances_cache.go:156] Start refreshing cloud provider node instances cache
I1003 22:39:55.477336       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 52
I1003 22:39:55.873546       1 node_instances_cache.go:168] Refresh cloud provider node instances cache finished, refresh took 396.27342ms
I1003 22:39:57.169224       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 52
I1003 22:39:57.173325       1 clusterstate.go:1094] Failed adding 1 nodes (1 unseen previously) to group vmss-<elided> due to OutOfResource.provisioning-state-failed; errorMessages=[]string{"Azure failed to provision a node for this node group"}
I1003 22:39:57.173740       1 static_autoscaler.go:854] Deleting 1 from vmss-<elided> node group because of create errors
I1003 22:39:57.173754       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 52
I1003 22:39:57.173794       1 azure_scale_set.go:376] Deleting vmss instances [azure:///subscriptions/<elided>/resourceGroups/<elided>/providers/Microsoft.Compute/virtualMachineScaleSets/vmss-<elided>/virtualMachines/679]
I1003 22:39:57.173867       1 azure_scale_set.go:426] Calling virtualMachineScaleSetsClient.DeleteInstancesAsync(&[679])
I1003 22:39:57.330951       1 static_autoscaler.go:439] Some nodes that failed to create were removed, skipping iteration
I1003 22:39:57.331031       1 azure_scale_set.go:184] Calling virtualMachineScaleSetsClient.WaitForDeleteInstancesResult(&[679]) for vmss-<elided>
```

</details>

At 22:40, the CAS logged this:
```
I1003 22:40:39.641141       1 azure_scale_set.go:149] VMSS: vmss-<elided>, returning in-memory size: 48
I1003 22:40:40.215536       1 orchestrator.go:295] Final scale-up plan: [{vmss-<elided> 48->53 (max: 1000)}]
I1003 22:40:40.215553       1 executor.go:147] Scale-up: setting group vmss-<elided> size to 53
```
At the time, the VMSS in question had 86 healthy instances, not 48. The graph below shows the CAS Ready/Unregistered stats, versus the number of nodes based on unique running kubelets. 5 seconds after the CAS scaled up the VMSS from its supposed size of 48 to 53, the VMSS began shrinking from 86 down to 53.

screenshot_013

After the scale-down finished, we started seeing occurrences of logs like this:

```
E1003 22:49:02.134021       1 azure_scale_set.go:191] virtualMachineScaleSetsClient.WaitForDeleteInstancesResult - DeleteInstances for instances &[679] for vmss-<elided> failed with error: Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
```

It really seems like the CAS is losing track of the instance count it's getting from Azure.

### PR DESCRIPTION
**Pull request by jackfrancis**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR adds a `StrictCacheUpdates` configuration for the Azure cluster-autoscaler provider (triggered by the `AZURE_STRICT_CACHE_UPDATES` runtime environment variable) that enables a modified Azure VMSS delete flow so that the local CA cache representation of node pool replica count is only updated upon a successful delete.

The current predictive cache update implementation was originally implemented here:

- https://github.com/kubernetes/autoscaler/pull/3036

The above works in most cases, but in situations where VMSS delete operations repeatedly fail, we see the the local CA cache gradually decrementing itself during each subsequent delete operation, which has the side-effect of eventually sending a VMSS replica update with a significantly reduced replica count when Azure API errors cease.

#### Which issue(s) this PR fixes:

Fixes #7432

#### Does this PR introduce a user-facing change?

```release-note
azure: StrictCacheUpdates to disable proactive vmss cache updates
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by comtalyst — COMMENTED** [review]
Seems like this is moving away from the existing idea of "proactively update size on deletion start, then invalidate cache if deletion fails" and going for "update size when deletion is complete".

I think both ideas have their own issues by nature, both from relying on an outdated assumption/prediction:
- Existing "proactively update" idea can have inconsistency on scale-up **while waiting** for an already-dead scale-down
  - Inconsistency **after** dead scale-down should be solved by refreshing cache on all layers
- New idea is more prone to scaling down below min count on scale-down **while waiting** for another scale-down to complete

Maybe incorporating the min count check heuristic that has been suggested in the issue might be one solution? Or sacrificing some performance and only allowing one scale-down operation at a time for a NodeGroup? 
If they work, I still hope both suggestions to be in the core than the provider in a long run. But if we are doing that in the provider for now, we should at least look at how to "reject" core calls properly without breaking state management.

<br/>
<br/>

> Existing "proactively update" idea can have inconsistency on scale-up **while waiting** for an already-dead scale-down

Still, I think we should try to understand the consequence of this more. Is the user-facing problem about overscaling, disruption (e.g., by "scale-up" operations removing nodes rather than adding due to new size being lower), etc? Is eventual consistency not good enough?

**Comment by jackfrancis** [comment]
@comtalyst Thanks for your review and perspective here.

Based on your analysis that there are trade-offs between both approaches, I think shipping a serialized implementation of cache updates after the Azure API has returned positive delete validation behind a configuration flag may help some customers who want to experiment with this alternate approach.

This change should have no effect on existing customers.

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set.go [hunk 1]
```diff
@@ -558,11 +561,18 @@ func (scaleSet *ScaleSet) waitForDeleteInstances(future *azure.Future, requiredI
...
+			}
+			scaleSet.invalidateInstanceCache()
+		}
 		return
 	}
-	// On failure, invalidate the instanceCache - cannot have instances in deletingState
-	scaleSet.invalidateInstanceCache()
+	if !scaleSet.manager.config.StrictCacheUpdates {
+		// On failure, invalidate the instanceCache - cannot have instances in deletingState
+		scaleSet.invalidateInstanceCache()
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_scale_set.go:574]: Should we add cache refresh here as well?
```suggestion
		if err := scaleSet.manager.forceRefresh(); err != nil {
				klog.Errorf("forceRefresh failed with error: %v", err)
		}
		scaleSet.invalidateInstanceCache()
```

**Review by comtalyst — APPROVED** [review]
Sounds like a fair mitigation, but let's still keep an eye on better approaches to solve both scale down below min count and delete failure inconsistency issue.

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set.go (hunk 1 above)
Comment (jackfrancis) [reply cluster-autoscaler/cloudprovider/azure/azure_scale_set.go:574]: Doing so would introduce a functional change for folks not opting into this new config feature. I think keeping the current behaviors unmodified for the default use case is a priority.

**Comment by comtalyst** [comment]
/lgtm
/approve
/hold

feel free to unhold whenever you are ready

**Pull request merged**